### PR TITLE
ProxySession cleanup: moving inline functions to .h

### PR DESCRIPTION
**Pull request by a-canary**
@shinrich request we move non-virtual, one-liner functions to the .h with "inline" keyword.

### CODE REVIEW

**Comment by a-canary** [comment]
Ready for review. @maskit @shinrich

**Comment by maskit** [comment]
I don't want to use `inline` unless it has huge difference in performance. Can you show me how much this improve performance?

Having implementations in headers brings dependencies in some cases, and it makes writing unit tests hard. As you know, we have a file for ugly stubs because of complex chained dependencies.

It also causes link issues actually. I don't remember what the function was, but just removing `inline` somewhere in iocore caused duplicate symbol error or something like that. Usually `inline` is just a hint for compilers but we depend on it somehow.

> move non-virtual, one-liner functions to the .h with "inline" keyword

This rule should be stated somewhere on docs if we did this. Of course with a consensus on the list.

**Comment by shinrich** [comment]
I think for trivial methods (e.g. simple class member accessors), you definitely don't want to incur the function call overhead.  If the function body is not in the header, I think it is highly unlikely that the compiler will pull it in.  But maybe that is not the case,  could be the compiler is smart enough to figure that out.  I'll try to run an experiment on that.  If the compiler can still inline trivial methods, I am fine with the declaration/definition separation.

I agree with @maskit on the inline keyword.  I normally just keep the implementation in the class, but @SolidWallOfCode showed me that if you have the implementation in the header, separate from the class definition you do need the inline keyword, at least with our current generation of gcc compilers.  Otherwise, you get the duplicate symbol errors.

**Comment by maskit** [comment]
I read some recent documents and articles about `inline`, and my conclusion is "No cheap tricks and let it be as compilers like". Because we require C++17, nobody should use ancient dumb compilers. I may not get the best executable, but it wouldn't be terrible.

https://en.cppreference.com/w/cpp/language/inline
https://pabloariasal.github.io/2019/02/28/cpp-inlining/

**Comment by a-canary** [comment]
@ywkaras was testing some of this yesterday. We were curious if g++ optimization would effectively inline code from one .cc to another .cc. If so then there is no reason to inline ever. 

Overall I would tend to agree with @maskit, we should avoid optimizing code unless it has a measurable, statistically significant performance increase. And even then there are strong advantages to keeping code human-readable for future iterations and maintenance.

**Comment by ywkaras** [comment]
This indicates that inlining does not happen if the body of a function appears in a different .cc file:
```
(conanrunenv) [root@d761696b8abf http2]# fgrep is_body_done *.cc *.h
Http2ConnectionState.cc:    if (len == 0 && !stream->is_body_done()) {
Http2ConnectionState.cc:  if (!stream->is_body_done() && payload_length == 0) {
Http2ConnectionState.cc:  if (stream->is_body_done() && !_sm->is_read_avail_more_than(0)) {
Http2Stream.cc:Http2Stream::is_body_done() const
Http2Stream.h:  bool is_body_done() const;
(conanrunenv) [root@d761696b8abf http2]# make V=1 Http2ConnectionState.o
depbase=`echo Http2ConnectionState.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
c++ -DHAVE_CONFIG_H -I. -I../../include  -Dlinux -D_LARGEFILE64_SOURCE=1 -D_COMPILE64BIT_SOURCE=1 -D_REENTRANT -D__STDC_LIMIT_MACROS=1 -D__STDC_FORMAT_MACROS=1 -I/Users/wkaras/REPOS/TS/iocore/eventsystem -I/Users/wkaras/REPOS/TS/iocore/net -I/Users/wkaras/REPOS/TS/iocore/net/quic -I/Users/wkaras/REPOS/TS/iocore/aio -I/Users/wkaras/REPOS/TS/iocore/hostdb -I/Users/wkaras/REPOS/TS/iocore/cache -I/Users/wkaras/REPOS/TS/iocore/utils -I/Users/wkaras/REPOS/TS/iocore/dns -I/Users/wkaras/REPOS/TS/include -I/Users/wkaras/REPOS/TS/lib -I/Users/wkaras/REPOS/TS/mgmt -I/Users/wkaras/REPOS/TS/mgmt/utils -I/Users/wkaras/REPOS/TS/proxy -I/Users/wkaras/REPOS/TS/proxy/http -I/Users/wkaras/REPOS/TS/proxy/hdrs -I/Users/wkaras/REPOS/TS/proxy/shared -I/Users/wkaras/REPOS/TS/proxy/http/remap  -D_GNU_SOURCE -I/opt/oath/openssl/1.1.1/include -DOPENSSL_NO_SSL_INTERN -I/opt/oath/brotli/1.0/include -std=c++17 -g -pipe -Wall -Wextra -Wno-ignored-qualifiers -Wno-unused-parameter -Wno-format-truncation -Wno-cast-function-type -Wno-stringop-overflow -O3 -feliminate-unused-debug-symbols -fno-strict-aliasing -Werror -Wno-invalid-offsetof -Wno-noexcept-type -mcx16  -MT Http2ConnectionState.o -MD -MP -MF $depbase.Tpo -c -o Http2ConnectionState.o Http2ConnectionState.cc &&\
mv -f $depbase.Tpo $depbase.Po
(conanrunenv) [root@d761696b8abf http2]# objdump --syms --demangle Http2ConnectionState.o | fgrep is_body_done
0000000000000000         *UND*	0000000000000000 Http2Stream::is_body_done() const
(conanrunenv) [root@d761696b8abf http2]#
```

From what I've heard, efforts to do inter-.cc-file inlining ( for example https://gcc.gnu.org/wiki/LinkTimeOptimization ) are still flakey.

For non-inlined small functions, gcc seems to do pretty well at optimizing out unnecessary stack operations as long as the optimization level is at least 2 (TS uses level 3):  https://godbolt.org/z/9Fx6CA

**Comment by a-canary** [comment]
@maskit do have you blocking concerns with this change. 

IMO, I don't care as long as the declaration and definition are separate. I would like the code organized so I can focus on macro optimizations. Inlining in the header file will not have worse performance, but it is unclear if it is faster. And I don't have time to do test performance metrics.

File: proxy/ProxySession.h [hunk 1]
```diff
@@ -181,3 +181,89 @@ class ProxySession : public VConnection
...
+///////////////////
+// INLINE
+
+static inline int64_t next_cs_id = 0;
+
+inline int64_t
+ProxySession::next_connection_id()
+{
+  return ink_atomic_increment(&next_cs_id, 1);
+}
```
Comment (ywkaras) [inline proxy/ProxySession.h]: Does more than one function use (the name) next_cs_id?  If not, why not
```
inline int64_t
ProxySession::next_connection_id()
{
  static int64_t next_cs_id = 0;
  return ink_atomic_increment(&next_cs_id, 1);
}
```
Pushing things to the lowest scope possible make future maintenance easier.

File: proxy/ProxyTransaction.h [hunk 2]
```diff
@@ -119,3 +116,98 @@ class ProxyTransaction : public VConnection
 
 private:
 };
+
+////////////////////////////////////////////////////////////
+// INLINE
+
+inline bool
+ProxyTransaction::is_transparent_passthrough_allowed()
```
Comment (ywkaras) [inline proxy/ProxyTransaction.h]: const?

**Comment by maskit** [comment]
I can live with definitions in header files, but want explicit rules on the doc. Otherwise people would put definitions randomly in the future.

Also, it's not only about performance. Definitions bring dependencies. Remember libUglyLogStubs.

File: proxy/ProxySession.h (hunk 1 above)
Comment (maskit) [reply proxy/ProxySession.h]: I don't think that is what he wants to do on this PR. You can make a PR for it.

**Comment by shinrich** [comment]
I agree with @maskit that this is a trade-off between performance and snow-balling dependencies.  In this particular case, I don't think the dependencies are an issue.

Where is the document on coding standards?  I can add something there.

**Comment by maskit** [comment]
Sorry for the late reply (I was away from hardware keyboards).

Here's the doc on coding standards.
https://cwiki.apache.org/confluence/display/TS/Coding+Style

**Comment by maskit** [comment]
Just for a note and I'm not sure at all, but Module that will be supported on C++20 might resolve some of this complication.
https://www.infoq.com/news/2019/07/cpp20-draft-finalized/

> Module support will be orthogonal to namespaces and will enable the structuring of large codebases into logical parts without requiring the separation of header and source files.

**Comment by zwoop** [comment]
Cherry-picked to v9.0.x branch.

**Pull request merged**